Allow override of checkout process

ksaandev opened the pull request:
<!-----------------------------------------------------------------------------
Thank you for contributing to the PrestaShop project! 

Please take the time to edit the "Answers" rows below with the necessary information.

Check out our contribution guidelines to find out how to complete it:
https://devdocs.prestashop.com/1.7/contribute/contribution-guidelines/#pull-requests
------------------------------------------------------------------------------>

To modify $this->checkoutProcess I need to be able to override it

| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | Move code to create `$this->checkoutProcess` into protected function to override it
| Type?         | bug fix
| Category?     | FO
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | https://github.com/PrestaShop/PrestaShop/issues/19848
| How to test?  | Use provided test module (see below). Install it and open Checkout page, you should see 2 "delivery steps" instead of one !

Test module - v2
[testmodule.zip](https://github.com/PrestaShop/PrestaShop/files/4931761/testmodule.zip)



<!-- Click the form's "Preview" button to make sure the table is functional in GitHub. Thank you! -->

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/prestashop/prestashop/19830)
<!-- Reviewable:end -->

ksaandev committed:
Allow override of checkout process

NeOMakinG commented:
Hey @ksaandev, looks like there is a problem with CSFixer, also, would it be possible that you create an issue for this ? It's necessary for the changelog and some things :/

Also, I think the QA needs to test it, maybe we need a module for them !

wdyt @PrestaShop/prestashop-core-developers

ksaandev committed:
Update OrderController.php

ksaandev commented:
> Hey @ksaandev, looks like there is a problem with CSFixer, also, would it be possible that you create an issue for this ? It's necessary for the changelog and some things :/
> 
> Also, I think the QA needs to test it, maybe we need a module for them !
> 
> wdyt @PrestaShop/prestashop-core-developers

Issue created https://github.com/PrestaShop/PrestaShop/issues/19848

Fixed PHP CS Fixer

ksaandev commented:
I cannot give you my module to test because it is paying.

Maybe I can make small module for demo

ksaandev committed:
Add hoo

matks commented:
Hi @ksaandev, PHP-CS-Fixer is not happy can you fix it ?

ksaandev committed:
Allow php-cs-fix

ksaandev commented:
> Hi @ksaandev, PHP-CS-Fixer is not happy can you fix it ?

Done !

matks commented on controllers/front/OrderController.php:
```diff
@@ -113,60 +113,11 @@ public function getCheckoutSession()
-                )
-            ));
+        $this->checkoutProcess = $this->buildCheckoutProcess($session, $translator);
+        
```
```suggestion

```

matks committed:
Remove unwanted blank space for php-cs-fixer

matks commented:
> > Hi @ksaandev, PHP-CS-Fixer is not happy can you fix it ?
> 
> Done !

Hi, I'm sorry 😅 php-cs-fixer is still red, there's an extra blank character. I will remove it 😉

matks commented:
@ksaandev PHP-CS-Fixer is happy now 😄 

I think this PR is valid, however we need
- to put this new hook Interesting into file install-dev/data/xml/hook.xml ; this is a registry for all new hooks
- insert your new hook into dedicated SQL table through the file install-dev/upgrade/sql/1.7.8.0.sql
- provide a small demo module to help QA test your PR

Please tell us if something is not clear in these needs 😉

ksaandev committed:
Update 1.7.8.0.sql

ksaandev committed:
Update hook.xml

ksaandev committed:
Allow override checkout process steps

ksaandev commented:
> @ksaandev PHP-CS-Fixer is happy now 😄
> 
> I think this PR is valid, however we need
> 
> * to put this new hook Interesting into file install-dev/data/xml/hook.xml ; this is a registry for all new hooks
> * insert your new hook into dedicated SQL table through the file install-dev/upgrade/sql/1.7.8.0.sql
> * provide a small demo module to help QA test your PR
> 
> Please tell us if something is not clear in these needs 😉

Hi ! I did all that and I added a test module in my 1st message in this PR

I also added a new function in CheckoutProcess to override steps which allows to insert new steps in middle of process (not only at the end), see how it works:
```
public function hookActionCheckoutRender($params)
    {
        $this->translator = Context::getContext()->getTranslator();
        /** @var CheckoutProcess $process */
        $process = $params['checkoutProcess'];
        $steps = $process->getSteps();

        $checkoutDeliveryStep = new CheckoutDeliveryStep(
            $this->context,
            $this->translator
        );
        $checkoutDeliveryStep->setCheckoutProcess($process);

        $newSteps = [
            $steps[0],
            $steps[1],
            $checkoutDeliveryStep,
            $steps[2],
            $steps[3]
        ];

        $process->overrideSteps($newSteps);
    }
```

ksaandev committed:
Merge branch 'develop' into patch-1

ksaandev commented:
Hi, I have also fix some git issues conflict

matks reviewed (dismissed):
This seems good, thank you @ksaandev. I approve this PR 👍

PierreRambaud commented on classes/checkout/CheckoutProcess.php:
```diff
@@ -108,6 +108,14 @@ public function getSteps()
+    /**
+     * @param array $steps
+     */
+    public function overrideSteps(array $steps)
```
```suggestion
    public function setSteps(array $steps)
```

PierreRambaud reviewed (changes requested):
As we use `addStep`, `getSteps` you must rename your method `setSteps` to be consistent.

PierreRambaud commented:
Only one thing and it's ok for me :)

ksaandev committed:
Improve CheckoutProcess setter name

Co-authored-by: GoT <PierreRambaud@users.noreply.github.com>

ksaandev commented:
@PierreRambaud thank you ! I applied your suggestion

SD1982 commented:
Hi @ksaandev I have an exception during the checkout process and it seems to be related to the provided module can you check this please ?
https://drive.google.com/file/d/1PWpffnZhQUkq9BlYn32Bkmwy0o48boZ8/view?usp=sharing

ksaandev commented:
> Hi @ksaandev I have an exception during the checkout process and it seems to be related to the provided module can you check this please ?
> https://drive.google.com/file/d/1PWpffnZhQUkq9BlYn32Bkmwy0o48boZ8/view?usp=sharing

Hi ! Yes you are right. Sorry about that, it's because we changed the code of the PR here https://github.com/PrestaShop/PrestaShop/pull/19830/commits/ddd0e697b37d1d6bef6e2842047102833406ec2f but not in my module. I will send you a new version of the module today

SD1982 commented:
Hi @ksaandev
Don't hesitate to ping me when you have updated the module and I will treat your PR as soon as it has been reviewed ! :+1:

ksaandev commented:
> Hi @ksaandev
> Don't hesitate to ping me when you have updated the module and I will treat your PR as soon as it has been reviewed ! 👍

I have updated my module to "Test module - v2" (see the PR description) to be synchronized with latest changes. Sorry for the error. I have checked on my shop and it works fine. Dont hesitate to call me if something bad

Thanks

SD1982 commented:
Hi @ksaandev it 's me again ! 
The exception seems to not appear anymore but the checkout process is locked on the third step 
When I click on the continue button the page is refreshed and nothing else happen 
https://drive.google.com/file/d/1Fvs9NhlmItzTp7KHo5buKWYdVtUz0bzc/view?usp=sharing

ksaandev commented:
> Hi @ksaandev it 's me again !
> The exception seems to not appear anymore but the checkout process is locked on the third step
> When I click on the continue button the page is refreshed and nothing else happen
> https://drive.google.com/file/d/1Fvs9NhlmItzTp7KHo5buKWYdVtUz0bzc/view?usp=sharing

Yes, it's because it's just to show it appears. It would need a big module to add a real step inside the process. The PR only adds the possibility for hook to add it, it is responsibility of module to then inject something good.

SD1982 commented:
Hello @ksaandev  Would it be possible to have a module that would allow us to check that the hook works correctly while allowing to go to the end of the command process?

PierreRambaud commented:
@SD1982 you can use https://github.com/PrestaShop/ps_qualityassurance and use the `actionCheckoutRender` hook :wink:

ksaandev commented:
Hi I am not sure https://github.com/PrestaShop/ps_qualityassurance can provide what @SD1982 wants.

Please try https://github.com/PrestaShop/ps_qualityassurance and tell me if it is good. If it is no good I will try make module for real checkout step but it is very long.

This PR only adds the step, it does not modify how checkout works. I am not sure module with complete step is necessary

SD1982 commented:
Hi @ksaandev  I have tested with https://github.com/PrestaShop/ps_qualityassurance  but i have a problem similar to your module i can"t go to the end of the checkout process . When the hook is triggered the page become blank and i can't continue the checkout 
https://drive.google.com/file/d/1og1YpGucYkDwlgNkQ6Wm3Dlu5LYSbbV5/view?usp=sharing

ksaandev commented:
> Hi @ksaandev I have tested with https://github.com/PrestaShop/ps_qualityassurance but i have a problem similar to your module i can"t go to the end of the checkout process . When the hook is triggered the page become blank and i can't continue the checkout
> https://drive.google.com/file/d/1og1YpGucYkDwlgNkQ6Wm3Dlu5LYSbbV5/view?usp=sharing

I know I said it 😄  https://github.com/PrestaShop/PrestaShop/pull/19830#issuecomment-662110890 

OK I will do it but it is going to take long time. I dont have lot of free time now maybe in some weeks.

SD1982 commented:
Hi @ksaandev 
Good news !! After some few additional tests and the help of @PierreRambaud and its [module](https://github.com/PrestaShop/ps_qualityassurance) I was able to correctly test your PR and correctly trigger the hook and go without error until the end of the command process .
So you don"t need to work to provide a new module !
Thanks @ksaandev !!

Progi1984 commented:
Thanks @ksaandev & @SD1982

ksaandev commented:
Thank you guys 😁
